providers/implementations/kem/hybrid_kem.c

```diff
@@ -0,0 +1,277 @@
+        if (ctx->ctxs[i] == NULL || !(*initf)(ctx->ctxs[i], auth->keys[i], params))
+            return 0;
+    }
+    return ossl_hybrid_set_ctx_params(ctx, params);
+}
+
+int ossl_hybrid_kem_auth_encapsulate_init(void *vctx, void *vkey,
+                                          void *vauthpriv,
+                                          const OSSL_PARAM params[])
+{
+    HYBRID_PKEY_CTX *ctx = (HYBRID_PKEY_CTX *)vctx;
+    HYBRID_PKEY *key = (HYBRID_PKEY *)vkey;
+    HYBRID_PKEY *authpriv = (HYBRID_PKEY *)vauthpriv;
+
+    return hybrid_kem_auth_init(ctx, key, authpriv, params,
+                                &EVP_PKEY_auth_encapsulate_init);
+}
+
+int ossl_hybrid_kem_auth_decapsulate_init(void *vctx, void *vkey,
+                                          void *vauthpub,
+                                          const OSSL_PARAM params[])
+{
+    HYBRID_PKEY_CTX *ctx = (HYBRID_PKEY_CTX *)vctx;
+    HYBRID_PKEY *key = (HYBRID_PKEY *)vkey;
+    HYBRID_PKEY *authpub = (HYBRID_PKEY *)vauthpub;
+
+    return hybrid_kem_auth_init(ctx, key, authpub, params,
+                                &EVP_PKEY_auth_decapsulate_init);
+}
+
+int ossl_hybrid_kem_encapsulate(void *vctx, unsigned char *enc, size_t *enclen,
+                                unsigned char *secret, size_t *secretlen)
+{
+    HYBRID_PKEY_CTX *ctx = (HYBRID_PKEY_CTX *)vctx;
+    size_t ciphertext_bytes, shared_secret_bytes;
+    unsigned int i;
+
+    if (ctx->pubkey_length == 0)
+        return 0;
+
+    if (secretlen == NULL || enclen == NULL) {
+        ERR_raise_data(ERR_LIB_PROV, ERR_R_PASSED_NULL_PARAMETER,
+                       "secretlen is NULL");
+        return 0;
+    }
+    if (enc == NULL) {
+        *enclen = ctx->ciphertext_bytes;
+        *secretlen = ctx->shared_secret_bytes;
+        return 1;
+    }
+    if (secret == NULL) {
+        ERR_raise_data(ERR_LIB_PROV, ERR_R_PASSED_NULL_PARAMETER,
+                       "secret is NULL");
+        return 0;
+    }
+    if (*secretlen < ctx->shared_secret_bytes) {
+        ERR_raise_data(ERR_LIB_PROV, PROV_R_BAD_LENGTH, "*secretlen too small");
+        return 0;
+    }
+    if (*enclen < ctx->ciphertext_bytes) {
+        ERR_raise_data(ERR_LIB_PROV, PROV_R_BAD_LENGTH, "*enclen too small");
+        return 0;
+    }
+
+    /* Encapsulate */
+    for (i = 0; i < ctx->info->num_algs; i++) {
+        ciphertext_bytes = ctx->info->alg[i].ciphertext_bytes;
+        shared_secret_bytes = ctx->info->alg[i].shared_secret_bytes;
+        if (!EVP_PKEY_encapsulate(ctx->ctxs[i], enc, &ciphertext_bytes,
+                                  secret, &shared_secret_bytes))
+            return 0;
+        if (ciphertext_bytes != ctx->info->alg[i].ciphertext_bytes
+                || shared_secret_bytes != ctx->info->alg[i].shared_secret_bytes) {
+            ERR_raise_data(ERR_LIB_PROV, PROV_R_BAD_LENGTH,
+                           "secretlen or enclen changed");
```

Comment:
Shouldn't it be 2 separate diagnostics?

Reply:
Given that this failure should never happen, I'm not sure how useful separate diagnostics would be.

If either of the conditions occurs it indicates a serious internal consistency problem.  Essentially, the algorithm lied about the sizes.

I can split this into two errors or, perhaps better, change it to an internal error if you insist.

Reply:
Yes, then internal error would be better

Reply:
I've done this too.